dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/DataFileCreator.java

```diff
@@ -38,489 +30,318 @@
-                    logger.warn("Exception while trying to uncompress file: " + fileName, ioe);
-                } finally {
-                    if (outputFile != null) {
-                        outputFile.delete();
-                    }
-                }
-            }
-        } else if (finalType.equalsIgnoreCase(ShapefileHandler.SHAPEFILE_FILE_TYPE)) {
-            // Shape files may have to be split into multiple files,
-            // one zip archive per each complete set of shape files:
+    private List<DataFile> unpackZipAndCreateDataFiles(Path tempFile, Long fileSizeLimit, Long zipFileUnpackFilesLimit) throws IOException {
+        List<DataFile> datafiles = new ArrayList<>();
 
-            //File rezipFolder = new File(this.getFilesTempDirectory());
-            File rezipFolder = getShapefileUnzipTempDirectory();
+        try (ZipInputStream unZippedIn = new ZipInputStream(Files.newInputStream(tempFile))) {
 
-            IngestServiceShapefileHelper shpIngestHelper;
-            shpIngestHelper = new IngestServiceShapefileHelper(tempFile.toFile(), rezipFolder);
+            ZipEntry zipEntry = unZippedIn.getNextEntry();
+            while (zipEntry != null) {
 
-            boolean didProcessWork = shpIngestHelper.processFile();
-            if (!(didProcessWork)) {
-                logger.error("Processing of zipped shapefile failed.");
-                return null;
-            }
+                // Note that some zip entries may be directories - we
+                // simply skip them:
 
-            try {
-                for (File finalFile : shpIngestHelper.getFinalRezippedFiles()) {
-                    FileInputStream finalFileInputStream = new FileInputStream(finalFile);
-                    finalType = determineContentType(finalFile);
-                    if (finalType == null) {
-                        logger.warn("Content type is null; but should default to 'MIME_TYPE_UNDETERMINED_DEFAULT'");
-                        continue;
-                    }
+                if (!zipEntry.isDirectory()) {
+                    logger.info("ZipEntry, file: " + zipEntry.getName());
 
-                    File unZippedShapeTempFile = saveInputStreamInTempFile(finalFileInputStream, fileSizeLimit);
-                    DataFile.ChecksumType checksumType = DataFile.ChecksumType.fromString(settingsService.getValueForKey(SettingsServiceBean.Key.FileFixityChecksumAlgorithm));
-                    DataFile new_datafile = createSingleDataFile(version, unZippedShapeTempFile, finalFile.getName(),
-                            finalType, checksumType, uncompressedSize);
-                    if (new_datafile != null) {
-                        datafiles.add(new_datafile);
-                    } else {
-                        logger.error("Could not add part of rezipped shapefile. new_datafile was null: " + finalFile.getName());
+                    if (datafiles.size() >= zipFileUnpackFilesLimit) {
```

Comment:
`zipFileUnpackFilesLimit > 0` - it is check if functionality is turned on
`datafiles.size() >= zipFileUnpackFilesLimit` - it is check if we already unpacked maximum amount of files that is enabled
These are the only conditions that checks for how much files we want to 
Since we do `datafiles.size() >= zipFileUnpackFilesLimit` before we try to create data file we will allow for one extra file to be unpacked.
Lets say limit is `100` and we already unpacked `100` files from zip.
This is the first time that this if will evaluate to true. So we are blocking file number `101` to be unpacked.
It seems ok to me.
Maybe you had in mind inconsistencies of how we treat some other limits (like limit of maximum file size)?

Reply:
>Maybe you had in mind inconsistencies of how we treat some other limits (like limit of maximum file size)?

Yes, but I mistakenly treated this `if` as if it was checking size of file and not the number of files.